src/libolm_migration.rs

```diff
@@ -0,0 +1,436 @@
+    /// Nullable so that a `PickledInboundGroupSession` can be constructed
+    /// incrementally. Must be populated!
+    #[wasm_bindgen(js_name = "roomId")]
+    pub room_id: Option<RoomId>,
+
+    /// Flag remembering if the session was directly sent to us by the sender
+    /// or if it was imported.
+    pub imported: bool,
+
+    /// Flag remembering if the session has been backed up.
+    #[wasm_bindgen(js_name = "backedUp")]
+    pub backed_up: bool,
+}
+
+#[wasm_bindgen]
+impl PickledInboundGroupSession {
+    /// Construct a new `PickledInboundGroupSession`, with default values.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[wasm_bindgen]
+impl Migration {
+    /// Migrate Megolm sessions of a libolm-based setup to a vodozemac-based setup
+    /// stored in an indexedDB crypto store.
+    ///
+    /// Before this method can be used, {@link #migrateBaseData} must be used to
+    /// import the base data into the crypto store.
+    ///
+    /// # Arguments
+    ///
+    /// * `sessions` - An `Array` of {@link PickledInboundGroupSession}s to import.
+    /// * `pickle_key` - The libolm pickle key that was used to pickle the megolm
+    ///   session objects.
+    /// * `store_handle` - A connection to the CryptoStore which will be used to
+    ///   store the vodozemac data.
+    #[wasm_bindgen(js_name = "migrateMegolmSessions")]
+    pub async fn migrate_megolm_sessions(
+        sessions: Array,
+        pickle_key: Uint8Array,
+        store_handle: &StoreHandle,
+    ) -> Result<JsValue, JsError> {
+        let rust_sessions = sessions
+            .into_iter()
+            .map(|s| libolm_pickled_megolm_session_to_rust_pickled_session(s, &(pickle_key.to_vec())))
+            .collect::<Result<_>>()?;
+
+        import_megolm_sessions_to_store(rust_sessions, store_handle.store.as_ref())
+            .await
+            .map_err(|e| JsError::from(&*e))?;
+        Ok(JsValue::UNDEFINED)
+    }
+}
+
+fn libolm_pickled_megolm_session_to_rust_pickled_session(
+    libolm_session: JsValue,
+    pickle_key: &[u8],
+) -> Result<matrix_sdk_crypto::olm::PickledInboundGroupSession> {
+    let libolm_session =
+        downcast::<PickledInboundGroupSession>(&libolm_session, "PickledInboundGroupSession")?;
+
+    let pickle = vodozemac::megolm::InboundGroupSession::from_libolm_pickle(
+        &libolm_session.pickle,
+        pickle_key,
+    )?
+    .pickle();
+
+    let sender_key = Curve25519PublicKey::from_base64(&libolm_session.sender_key)?;
+
+    Ok(matrix_sdk_crypto::olm::PickledInboundGroupSession {
+        pickle,
+        sender_key,
+        signing_key: iter::once((
+            DeviceKeyAlgorithm::Ed25519,
+            Ed25519PublicKey::from_base64(&libolm_session.sender_signing_key)?.into(),
+        ))
+        .collect(),
+        room_id: libolm_session
+            .room_id
+            .clone()
+            .ok_or(JsError::new("Room ID not specified for megolm session"))?
```

Comment:
`.context` here too

Reply:
this method returns `Result<_, JsError>`, rather than `std::Error` or whatever, so I don't think `.context` works here. Have switched to `.ok_or_else` though.